### PR TITLE
Refactor/ELEC467

### PR DESCRIPTION
**Pull request by ahnaftazwar368**
This will be a fun PR to read :').

- Once user services and report services are also moved to prisma, one more PR is needed to just remove typeORM leftovers
- Removal of database folder and associated docker parts once prisma migration application is sorted

### CODE REVIEW

File: backend/Dockerfile [hunk 1]
```diff
@@ -5,12 +5,15 @@ FROM node:20.2-alpine as builder
 WORKDIR /app
 
 COPY package.json package-lock.json tsconfig.json ./
+COPY prisma ./prisma
 RUN npm install
+# RUN npx prisma generate
```
Comment (lhjt) [inline backend/Dockerfile:10]: Why is prisma generate not being run in the build stage?

**Review by lhjt — CHANGES_REQUESTED** [review]

File: backend/Dockerfile (hunk 1 above)
Comment (ahnaftazwar368) [reply backend/Dockerfile]: I've updated it to run in both stages now, I thought it might not be necessary during the build stage so I had commented it out. Is this fine now? 🤔

Comment (lhjt) [reply backend/Dockerfile]: Is it required to actually build the app? I just assumed it was, hence I asked about it (as prisma injects typescript information from my understanding).

Comment (lhjt) [reply backend/Dockerfile]: image
Oh nvm it's because prisma isn't being used properly gg; please address this before fully converting this project to prisma ‼️

Comment (ahnaftazwar368) [reply backend/Dockerfile]: Nah i dont think the container uses the prisma generate to build first?? (it works fine without it)

yup the code gets quite problematic without the rawquery for courses. Pure prisma should be super doable for all other services though. 
Unless you can see a cool way around it? 👀

Comment (lhjt) [reply backend/Dockerfile]: I don't extensively use prisma so I'm not sure if there's a pure prisma way to do what you're trying to achieve, unfortunately - do you see pure prisma being used later down the line for other things?

Comment (ahnaftazwar368) [reply backend/Dockerfile]: yup, all other services will be pure prisma, i tried to make course pure prisma too but n^26 time complexity wasnt worth it 💀

**Review by lhjt — APPROVED** [review]

File: backend/prisma/schema.prisma [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  overallRating     Float       @map("overall_rating")
+  reportes           reports[]
+  courses           courses     @relation(fields: [courseCode], references: [courseCode], onDelete: NoAction, onUpdate: NoAction, map: "fk_course_code")
+  users             users       @relation(fields: [zid], references: [zid], onDelete: NoAction, onUpdate: NoAction, map: "fk_zid")
+}
+
+model users {
+  zid                String      @id(map: "pk_zid")
+  bookmarkedReviews String[]     @map("bookmarked_reviews")
+  bookmarkedCourses String[]     @map("bookmarked_courses")
```
Comment (ahnaftazwar368) [inline backend/prisma/schema.prisma:76]: unfortunate design decision. These need to be changed in the future to actually be the reviews and courses.

File: migration/package.json [hunk 3]
```diff
@@ -1,6 +1,7 @@
 {
   "name": "migration",
   "version": "1.0.0",
+  "type": "module",
```
Comment (aimenhamed) [inline migration/package.json:4]: why was this added?

File: backend/Dockerfile [hunk 4]
```diff
@@ -21,7 +24,10 @@ COPY package.json package-lock.json ./
 RUN npm install --production
 
 COPY --from=builder /app/dist ./dist
+COPY --from=builder /app/config ./config
```
Comment (aimenhamed) [inline backend/Dockerfile:27]: why are we adding config since we removed to used envsafe?

File: backend/package.json [hunk 5]
```diff
@@ -8,19 +8,21 @@
...
   },
   "dependencies": {
+    "@prisma/client": "^4.15.0",
     "cors": "^2.8.5",
     "envsafe": "^2.0.3",
     "express": "^4.18.2",
     "ioredis": "^5.3.2",
     "jsonwebtoken": "^9.0.0",
     "pg": "^8.10.0",
+    "prisma": "^4.15.0",
```
Comment (aimenhamed) [inline backend/package.json:25]: I think this is a dev dependency

File: backend/prisma/schema.prisma [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  faculty            String
+  fieldOfEducation   String     @map("field_of_education")
+  genEd              Boolean     @map("gen_ed")
+  level              Int
+  school             String
+  studyLevel         String     @map("study_level")
+  terms              Int[]
+  title              String
+  uoc                Int
+  rating             Float
```
Comment (aimenhamed) [inline backend/prisma/schema.prisma:36]: lets remove this field

File: backend/prisma/schema.prisma [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  grade             Int?
+  termTaken        String      @map("term_taken")
+  createdTimestamp  DateTime    @default(now()) @db.Timestamp(6) @map("created_timestamp")
+  updatedTimestamp  DateTime    @default(now()) @db.Timestamp(6) @map("updated_timestamp")
+  upvotes           String[]
+  manageability     Float
+  usefulness        Float
+  enjoyability      Float
+  overallRating     Float       @map("overall_rating")
+  reportes           reports[]
```
Comment (aimenhamed) [inline backend/prisma/schema.prisma:68]: typo ?

File: backend/prisma/schema.prisma [hunk 8]
```diff
@@ -0,0 +1,80 @@
+generator client {
```
Comment (aimenhamed) [inline backend/prisma/schema.prisma:1]: might be good opportunity to re assess our tables and design properly

File: backend/src/api/schemas/report.schema.ts [hunk 9]
```diff
@@ -28,7 +28,7 @@ export type UpdateReportStatus = z.infer<typeof UpdateReportStatusSchema>;
 export const ReportSchema = z
   .object({
     reportId: z.string(),
-    review: ReviewSchema,
+    reviews: ReviewSchema,
```
Comment (aimenhamed) [inline backend/src/api/schemas/report.schema.ts:31]: not sure whats happening here

File: backend/src/env.ts [hunk 10]
```diff
@@ -29,13 +29,17 @@ export const env = envsafe({
     devDefault: "localhost",
   }),
   REDIS_USER: str({
-    allowEmpty: process.env.NODE_ENV !== "production",
+    allowEmpty: true,
```
Comment (aimenhamed) [inline backend/src/env.ts:32]: why ?

File: backend/src/modules/prisma.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+import { getLogger } from "../utils/logger";
+import env from "../env";
+
+export default class PrismaClient {
+  private logger = getLogger();
+  private connection: Connection;
+
+  constructor() {
+    this.connection = new Connection({
+      // log: ["query"],
```
Comment (aimenhamed) [inline backend/src/modules/prisma.ts:11]: would be good for debugging to get logs of the output sql queries when locally developing

File: backend/src/repositories/review.repository.ts [hunk 12]
```diff
@@ -1,42 +1,60 @@
...
+    return await this.prisma.reviews.findMany();
   }
 
-  async getCourseReviews(courseCode: string): Promise<ReviewEntity[]> {
-    return await this.manager.find(ReviewEntity, {
+  async getCourseReviews(courseCode: string): Promise<reviews[]> {
+    return await this.prisma.reviews.findMany({
       where: {
-        courseCode,
+        courseCode: courseCode,
```
Comment (aimenhamed) [inline backend/src/repositories/review.repository.ts:14]: O_o

**Review by aimenhamed — CHANGES_REQUESTED** [review]

File: backend/Dockerfile (hunk 4 above)
Comment (ahnaftazwar368) [reply backend/Dockerfile:27]: looks like leftover code form merging the envsafe change into the branch, i shall remove 🙏

File: backend/package.json (hunk 5 above)
Comment (ahnaftazwar368) [reply backend/package.json:25]: donezo, i think we spotted this issue before but looks like i forgot to move it down 😭

File: backend/prisma/schema.prisma (hunk 6 above)
Comment (ahnaftazwar368) [reply backend/prisma/schema.prisma:36]: would that be fine for the database? I wasn't sure if it would break all the old data or reviews so i didnt touch the actual schema

File: backend/prisma/schema.prisma (hunk 7 above)
Comment (ahnaftazwar368) [reply backend/prisma/schema.prisma:68]: we dont talk about it

File: backend/src/modules/prisma.ts (hunk 11 above)
Comment (ahnaftazwar368) [reply backend/src/modules/prisma.ts:11]: yes but surely we dont need permanent debugging no??

File: backend/src/repositories/review.repository.ts (hunk 12 above)
Comment (ahnaftazwar368) [reply backend/src/repositories/review.repository.ts:14]: bro pleaseeeeee 😭😭😭 desperate debugging calls for desperate measures, pretend u didnt see this

File: migration/package.json (hunk 3 above)
Comment (ahnaftazwar368) [reply migration/package.json:4]: app.ts is recognised as an es module and so importing it into index.ts doesnt work with commonjs, docker loses its mind without the module type definition. Perhaps you have a different solution?
Screenshot 2023-07-16 at 10 46 36 am

File: backend/src/env.ts (hunk 10 above)
Comment (ahnaftazwar368) [reply backend/src/env.ts:32]: failsafe, envsafe will trigger for when we startup our production build and ask for login. is it not true that we do not want to have credentials for redis in more or less all cases?
Screenshot 2023-07-16 at 10 54 08 am

File: backend/prisma/schema.prisma (hunk 6 above)
Comment (aimenhamed) [reply backend/prisma/schema.prisma:36]: only database we have right now is the staging one, which is fine for us to delete and redo again

File: backend/src/env.ts (hunk 10 above)
Comment (aimenhamed) [reply backend/src/env.ts:32]: we dont have values for it in all case (local)

File: backend/src/modules/prisma.ts (hunk 11 above)
Comment (aimenhamed) [reply backend/src/modules/prisma.ts:11]: we dont want prod logs for this

File: migration/package.json (hunk 3 above)
Comment (aimenhamed) [reply migration/package.json:4]: okies congratulations or sorry to hear that

File: backend/src/env.ts (hunk 10 above)
Comment (ahnaftazwar368) [reply backend/src/env.ts:32]: doesnt seem like we should need credentials for redis now so i reckon we make sure it never bothers us till that changes

File: backend/src/modules/prisma.ts (hunk 11 above)
Comment (ahnaftazwar368) [reply backend/src/modules/prisma.ts:11]: hmm im not too sure what ur saying then 🤔 what exactly do you think should get logged

File: backend/prisma/schema.prisma (hunk 6 above)
Comment (ahnaftazwar368) [reply backend/prisma/schema.prisma:36]: yeah so just removing the rating field means migrations from migration server doesnt work.

Also now that i have a closer look, migration server is still hard coded up with typeORM 😭, i think we'll need to remove the orm from the migration server so we can start flexibly changing the db structure without pain.

File: backend/src/modules/prisma.ts (hunk 11 above)
Comment (aimenhamed) [reply backend/src/modules/prisma.ts:11]: `log: process.env.NODE_ENV !== "production" && ["query"]`

**Review by aimenhamed — APPROVED** [review]

**Review by aimenhamed — APPROVED** [review]
